cumulus/tasks/pdr-status-check/index.js

```diff
@@ -1,123 +1,177 @@
-
-          eventToCheck.payload.counter = counter + 1;
-          eventToCheck.payload.limit = limit;
-          eventToCheck.payload.running = running;
-          eventToCheck.payload.completed = completed;
-          eventToCheck.payload.failed = failed;
-        }
-
-        return Promise.resolve(cb(null, eventToCheck));
-      });
-    }
-
-    return Promise.all(runningExecutions.map(updateStatus))
-      .catch(handleError);
+/**
+ * Create a task return value for a set of executions
+ *
+ * Example output:
+ *   {
+ *     isFinished: false,
+ *     running: ['arn:123'],
+ *     failed: [
+ *       { arn: 'arn:456', reason: 'Workflow Aborted' }
+ *     ],
+ *     completed: []
+ *   }
+ *
+ * @param {Object} event - the event that came into checkPdrStatuses
+ * @param {Object} groupedExecutions - a map of execution statuses grouped
+ *   by status
+ * @returns {Object} - a description of the results of this task execution
+ */
+function buildOutput(event, groupedExecutions) {
+  // eslint-disable-next-line require-jsdoc
+  const getExecutionArn = (execution) => execution.executionArn;
+
+  // eslint-disable-next-line require-jsdoc
+  const parseFailedExecution = (execution) => {
+    let reason = 'Workflow Failed';
+    if (execution.output) reason = JSON.parse(execution.output).exception;
+    return { arn: execution.executionArn, reason };
+  };
+
+  // eslint-disable-next-line require-jsdoc
+  const parseAbortedExecution = (execution) => {
+    let reason = 'Workflow Aborted';
+    if (execution.output) reason = JSON.parse(execution.output).exception;
+    return { arn: execution.executionArn, reason };
+  };
+
+  const running = groupedExecutions.running.map(getExecutionArn);
+
+  const failed = (event.input.failed || [])
+    .concat(groupedExecutions.aborted.map(parseAbortedExecution))
+    .concat(groupedExecutions.failed.map(parseFailedExecution));
+
+  const completed = (event.input.completed || [])
+    .concat(groupedExecutions.completed.map(getExecutionArn));
+
+  const output = {
+    isFinished: groupedExecutions.running.length === 0,
+    running,
+    failed,
+    completed
+  };
+
+  if (!output.isFinished) {
+    output.counter = getCounterFromEvent(event) + 1;
+    output.limit = getLimitFromEvent(event);
   }
 
-  return checkStatus(event).catch(handleError);
-};
+  return output;
+}
+
+/**
+ * Checks a list of Step Function Executions to see if they are all in
+ * terminal states.
+ *
+ * @param {Object} event - a Cumulus Message that has been sent through the
+ *   Cumulus Message Adapter
+ * @returns {Promise.<Object>} - an object describing the status of Step
+ *   Function executions related to a PDR
+ */
+function checkPdrStatuses(event) {
+  const runningExecutionArns = event.input.running || [];
+
+  const promisedExecutionDescriptions = runningExecutionArns.map((executionArn) =>
+    aws.sfn().describeExecution({ executionArn }).promise());
+
+  return Promise.all(promisedExecutionDescriptions)
```

Comment:
This might hit the step function api limit if there are too many executions. In case of aster there are about 1000 granules per PDR which could easily reach the step function api limit.

I usually use this library to limit the number of concurrent executions: https://www.npmjs.com/package/p-limit

Reply:
Agreed, I'll file a ticket to address that.  (It was a potential issue before this PR as well, so seems like it should be a new ticket.)

Reply:
Created [CUMULUS-304](https://bugs.earthdata.nasa.gov/browse/CUMULUS-304) to address this.